Python/chapter02/2.4 - Partition/miguel_2.4_soln.py

```diff
@@ -0,0 +1,276 @@
+"""
+Python version 3.7.0
+2.4 - Partition
+Write code to partition a linked list around a value x,
+such that all nodes less than x come before all nodes
+greater than or equal to x. If x is contained within
+the list, the values of x only need to be after
+the elements less than x. The partition element x
+can appear anywhere in the 'right partition'; it does
+not need to appear between the left and right
+partitions.
+EXAMPLE
+Input:  3 -> 5 -> 8 -> 5 -> 10 -> 2 -> 1 [pivot = 5]
+Result: 3 -> 1 -> 2 -> 10 -> 5 -> 5 -> 8
+"""
+import unittest
+from typing import Optional
+
+
+class Node:
+    def __init__(self, d: int):
+        self.data = d
+        self.next = None
+
+    def __repr__(self):
+        return self.__str__()
+
+    def __str__(self):
+        return '<Node Value: {}>'.format(self.data)
+
+    def __eq__(self, other: object):
+        if not isinstance(other, Node):
+            return NotImplemented
+        return self.data == other.data
+
+
+class LinkedList:
+    def __init__(self, *numbers: int):
+        self.head = None
+        self.tail = None
+        self.size = 0
+        for num in numbers:
+            self.append_to_tail(num)
+
+    def append_to_tail(self, e) -> None:
+        if isinstance(e, int):
+            self._append_num(e)
+        elif isinstance(e, Node) or e is None:
+            self._append_node(e)
+
+    def _append_num(self, d: int) -> None:
+        if self.head is None:
+            self.head = Node(d)
+            self.tail = self.head
+        else:
+            end = Node(d)
+            self.tail.next = end
+            self.tail = end
+        self.size += 1
+
+    def _append_node(self, n: Optional[Node] = None) -> None:
+        if self.head is None:
+            self.head = n
+            self.tail = self.head
+        else:
+            end = n
+            self.tail.next = end
+            self.tail = end
+        self.size += 1
+
+    def append_to_head(self, d: int) -> None:
+        new_head = Node(d)
+        new_head.next = self.head
+        self.head = new_head
+        self.size += 1
+
+    def get_node_at(self, index: int) -> Node:
+        if index < 0 or index >= self.size:
+            raise IndexError('list index out of range')
+        n = self.head
+        for i in range(self.size):
+            if i == index:
+                return n
+            n = n.next
+
+    def pop_head(self) -> Optional[Node]:
+        if self.head is None:
+            return self.head
```

Comment:
maybe raise something instead of returning `None`. it's kinda like doing `[].pop()`

Reply:
doesn't that defeat the purpose of having Optional[Node]?

Also, for `right_partition.append_to_tail(ll.pop_head())` this leads to an error if I don't return None.  By that line of code, the original linked list `ll` no longer has a head.

Reply:
Disregard my comment, found out that I don't need that line of code after the while loop anymore lol

Reply:
that totally defeats the purpose of having it return `Optional[Node]`, and that's a good thing! ultimately, why do you _need_ to return that? isn't it better to make it behave like `list.pop()`? otherwise you're going against the [principle of least astonishment](https://en.m.wikipedia.org/wiki/Principle_of_least_astonishment).

same thing for the latter: why do you _need_ that line? isn't it better to fix that problem elsewhere within this class to avoid the implementation details from leaking?

Reply:
updated 👍